rfcs/0000-arbitrary-multisig/0000-arbitrary-multisig.md

```diff
@@ -0,0 +1,354 @@
+
+### Signature Packet Format 2
+
+Signature Packat Format 1 contains the following data concatenated together:
+
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 2 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 20 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* witness source in 64-bit little-endian unsigned integer
+* witness index in 32-bit little-endian unsigned integer
+* witness offset in 32-bit little-endian unsigned integer
+* witness length in 32-bit little-endian unsigned integer
+
+Note that pipe is capable of accepting binary data with zeros, the data above do not need zero escape encoding.
+
+In this case, the leaf script also locates a slice of data based on the 4 witness locating variables. Different from the previous version, the leaf script expects the located slice of data to be of the following format:
+
+* Public key of a given digital signature scheme.
+* Signature of a given digital signature scheme.
+
+No trailing data should be present after the signature.
+
+The digital signature scheme to use here, must be pre-determined for a leaf script. The leaf script then verifies the signature against the public key and the signing message in `argv[0]`.
+
+When the signature verification process fails, the leaf script sends the following response packet via `leaf_to_root_pipe[1]` to the root script:
+
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 1 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+
+The leaf script then terminates with a non-zero exit code.
+
+If all verification work succeeds, the leaf script sends the following response packet via `leaf_to_root_pipe[1]` to the root script:
+
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+* 0 encoded in [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) format.
+
+The leaf script then waits to read more packets for verification from `root_to_leaf_pipe[0]`. There is no situation when a leaf script terminates with a success exit code. We rely on CKB's behavior that when the root VM terminates, all the child VMs also terminate.
+
+This signature packet format is in fact designed for `ALGO_ID` of `63`, when an arbitrary script is loaded for signature verification. Signature packet format 2 can be used to exchange signatures to valid between root script and the arbitrary leaf script.
+
+## Exec
+
+A root script might use exec syscall to invoke a leaf script, when the following conditions are met:
```

Comment:
If there is a security issue or malicious code in the leaf script executed via `exec`, an early `exit(0)` can bypass all validation. It should be used with extreme caution.

For example, Alice, Bob, and Charlie agree to use a 2-of-3 signature scheme, but Charlie is aware of a bug in a leaf script that allows `exit(0)` to bypass validation. As a result, he can exploit it to proceed with only one signature.

Reply:
For one thing, a multisig implementation is fine that it just support spawn syscalls.

To really discuss the code here, I think this is too strong a convention:

```
Charlie is aware of a bug in a leaf script that allows exit(0) to bypass validation
```

Exactly how will this bug proceed? 

To me, it could also happens that Charlie is aware of a bug in a root script that turns out to be `exit(0)`. In this case Charlie just exploits the root script. The fact that the return code of a leaf script becomes the return code of the root script, does not always mean one should always assume the worst.

What if I assume the following:

```
Charlie finds a way that a leaf script sends all zeros via the response pipe.
```

How can we tell one is easier to achieve than the other?

Reply:
Maybe we could include a chapter on security, explaining various scenarios.

Reply:
Done